README.md
# Trip Sorter

[![Build Status](https://travis-ci.org/sasakocic/trip-sorter.svg)](https://travis-ci.org/sasakocic/trip-sorter)
[![Code Style Status](https://github.styleci.io/repos/101542570/shield?branch=master)](https://styleci.io/repos/101542570)
[![CodeCov](https://img.shields.io/codecov/c/github/sasakocic/trip-sorter.svg)](https://codecov.io/gh/sasakocic/trip-sorter)
[![CodeClimate](https://api.codeclimate.com/v1/badges/024ca63cc345a23bf71d/maintainability.svg)](https://codeclimate.com/github/sasakocic/trip-sorter)
[![Issue Count](https://codeclimate.com/github/sasakocic/trip-sorter/badges/issue_count.svg)](https://codeclimate.com/github/sasakocic/trip-sorter)
[![Issue Count](https://scrutinizer-ci.com/g/sasakocic/trip-sorter/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/sasakocic/trip-sorter/?branch=master)

## Task

You are given a stack of boarding cards for various transportations that will take you from a point A to point B via several stops on the way. All of the boarding cards are out of order, and you don't know where your journey starts, nor where it ends. Each boarding card contains information about seat assignment, and means of transportation (such as flight number, bus number etc).
Write an API that lets you sort this kind of list and present back a description of how to complete your journey.
For instance the API should be able to take an unordered set of boarding cards, provided in a format defined by you, and produce this list:

1. Take the train 78A from Madrid to Barcelona. Sit in the seat 45B.
2. Take the airport bus from Barcelona to Gerona Airport. No seat assignment.
3. From Gerona Airport, take flight SK455 to Stockholm. Gate 45B, seat 3A. Baggage drop at ticket counter 344.
4. From Stockholm, take flight SK22 to New York JFK. Gate 22, seat 7B. Baggage will we automatically transferred from your last leg.
5. You have arrived at your final destination.

The list should be defined in a format that's compatible with the input format.
The API is to be an internal PHP API so it will only communicate with other parts of a PHP application, not server to server, nor server to client.
Use PHP-doc to document the input and output your API accepts/returns.

## Assumptions

- Since the result is only one list, I assume that input receives boarding cards for just one particular trip, and we just need to sort them.
- To cover all variations of output without introducing artificial intelligence, I will assume that cards have `Sit in seat 45B.` or `No seat assignment.` written on them. Having a field seat would be considered otherwise.

## Installation

Install with [Composer](https://getcomposer.org/)

    composer install

## Tests

Tests can be executed by running the following command from the root directory

    composer run tests

## Usage

Input is an array in the following format
```php
$cards = [
  ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in the seat 45B.'],
  ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
  ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'seat' => '3A', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
  ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'seat' => '7B', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
];
$result = TripSorter::sort($cards);
```

Result is an array of strings:

1. Take train 78A from Madrid to Barcelona. Sit in seat 45B.
2. Take the airport bus from Barcelona to Gerona Airport. No seat assignment.
3. From Gerona Airport, take flight SK455 to Stockholm. Gate 45B, seat 3A.
Baggage drop at ticket counter 344.
4. From Stockholm, take flight SK22 to New York JFK. Gate 22, seat 7B.
Baggage will we automatically transferred from your last leg.
5. You have arrived at your final destination.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.


## License

Please see the [license file](LICENSE) for more information.


CHANGELOG.md
# CHANGELOG

We follow [SemVer](http://semver.org/).

## 2.0 (2020-04-18)
- refactor of classes for simplified architecture
- added helper
- separated printer

## 1.1 (2020-04-06)
- refactored namespace
- composer.json is updated
- edited phpunit.xml
- updated PSR-2 to PSR-12

## 1.0 (2017-08-28)
 - Initial release.


tests/TripSorterTest.php
<?php

namespace Test;

use Skocic\TripSorter\TripHelper;
use Skocic\TripSorter\TripPrinter;
use Skocic\TripSorter\TripSorter;
use PHPUnit\Framework\TestCase;

class TripSorterTest extends TestCase
{
    /** @var TripSorter */
    private TripSorter $trip;
    private TripPrinter $printer;

    /**
     * Setup.
     */
    final public function setUp(): void
    {
        $this->trip = new TripSorter();
        $this->printer = new TripPrinter();
    }

    /**
     * Test do method.
     */
    final public function testDoWithEmpty(): void
    {
        $result = $this->trip->sort();
        $this->assertSame(['You have arrived at your final destination.'], ($this->printer)($result));
    }

    /**
     * Test sorting no cards.
     */
    final public function testSortEmpty(): void
    {
        $cards = [];
        $result = $this->trip->sort($cards);
        $this->assertSame([], $result);
    }

    /**
     * Test with only one card.
     */
    final public function testSortingOneCard(): void
    {
        $cards = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
        ];
        $boardingCards = TripHelper::toBoardingCards($cards);
        $afterSort = $this->trip->sort($boardingCards);
        $this->assertSame($boardingCards, $afterSort);
    }

    /**
     * Test alread sorted cards.
     */
    final public function testSorted(): void
    {
        $cards = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $expected = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $result = $this->trip->sort(TripHelper::toBoardingCards($cards));
        $this->assertSame($expected, TripHelper::toArray($result));
    }

    /**
     * Test unsorted cards.
     */
    public function testUnsorted()
    {
        $cards = TripHelper::toBoardingCards([
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
        ]);
        $expected = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $result = $this->trip->sort($cards);
        $this->assertSame($expected, TripHelper::toArray($result));
    }

    public function testWithAmbigiousStart()
    {
        $cards = TripHelper::toBoardingCards([
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
        ]);
        $expected = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $this->expectExceptionMessage('There should be one starting point');
        $result = $this->trip->sort($cards);
    }
}


tests/TripPrinterTest.php
<?php

namespace Test;

use Skocic\TripSorter\TripHelper;
use Skocic\TripSorter\TripPrinter;
use Skocic\TripSorter\TripSorter;
use PHPUnit\Framework\TestCase;

class TripPrinterTest extends TestCase
{
    /** @var TripSorter */
    private TripSorter $trip;
    private TripPrinter $printer;

    /**
     * Setup.
     */
    final public function setUp(): void
    {
        $this->trip = new TripSorter();
        $this->printer = new TripPrinter();
    }

    public function testOutput(): void
    {
        $cards = TripHelper::toBoardingCards([
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ]);
        $expected = [
            'Take train 78A from Madrid to Barcelona. Sit in seat 45B.',
            'Take airport bus  from Barcelona to Gerona Airport. No seat assignment.',
            'From Gerona Airport, take flight SK455 to Stockholm. Gate 45B, seat 3A. Baggage drop at ticket counter 344.',
            'From Stockholm, take flight SK455 to New York JFK. Gate 22. Baggage will we automatically transferred from your last leg.',
            'You have arrived at your final destination.',
        ];
        $result = $this->trip->sort($cards);
        $this->assertSame($expected, ($this->printer)($result));
    }
}


tests/Tests/TripSorterTest.php
<?php

namespace Tests;

use PHPUnit\Framework\TestCase;
use App\TripSorter;

class TripSorterTest extends TestCase
{
    /** @var TripSorter */
    private $service;

    /**
     * Setup service.
     */
    public function setUp()
    {
        $this->service = new TripSorter();
    }

    /**
     * Test do method.
     */
    public function testDoWithEmpty()
    {
        $result = TripSorter::sort();
        $this->assertSame(['You have arrived at your final destination.'], $result);
    }

    /**
     * Test sorting no cards.
     */
    public function testSortEmpty()
    {
        $cards = [];
        $result = $this->service->setCards($cards)->sortCards()->getCards();
        $this->assertSame([], $result);
    }

    /**
     * Test with only one card.
     */
    public function testSortingOneCard()
    {
        $cards = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
        ];
        $beforeSort = $this->service->setCards($cards)->getCards();
        $afterSort = $this->service->sortCards()->getCards();
        $this->assertSame($beforeSort, $afterSort);
    }

    /**
     * Test alread sorted cards.
     */
    public function testSorted()
    {
        $cards = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $expected = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $result = $this->service->setCards($cards)->sortCards()->getCards();
        $this->assertSame($expected, $result);
    }

    /**
     * Test unsorted cards.
     */
    public function testUnsorted()
    {
        $cards = [
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
        ];
        $expected = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $result = $this->service->setCards($cards)->sortCards()->getCards();
        $this->assertSame($expected, $result);
    }

    /**
     * @expectedException \RuntimeException
     * @expectedExceptionMessage There should be one starting point
     */
    public function testWithAmbigiousStart()
    {
        $cards = [
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
        ];
        $expected = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $result = $this->service->setCards($cards)->sortCards()->getCards();
        $this->assertSame($expected, $result);
    }

    public function testOutput()
    {
        $cards = [
            ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
            ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
            ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
            ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
        ];
        $expected = [
            'Take train 78A from Madrid to Barcelona. Sit in seat 45B.',
            'Take airport bus  from Barcelona to Gerona Airport. No seat assignment.',
            'From Gerona Airport, take flight SK455 to Stockholm. Gate 45B, seat 3A. Baggage drop at ticket counter 344.',
            'From Stockholm, take flight SK455 to New York JFK. Gate 22. Baggage will we automatically transferred from your last leg.',
            'You have arrived at your final destination.',
        ];
        $result = $this->service->setCards($cards)->sortCards()->output();
        $this->assertSame($expected, $result);
    }
}


src/TripPrinter.php
<?php

namespace Skocic\TripSorter;

class TripPrinter
{
    /**
     * @param BoardingCard[] $cards
     * @return array
     */
    public function __invoke(array $cards): array
    {
        $output = [];
        foreach ($cards as $card) {
            if ($card->getType() === 'flight') {
                $output[] = 'From ' . $card->getFrom()
                    . ', take ' . $card->getType()
                    . ' ' . $card->getNumber()
                    . ' to ' . $card->getTo()
                    . '. ' . $card->getInfo();
            } else {
                $output[] = 'Take ' . $card->getType()
                    . ' ' . $card->getNumber()
                    . ' from ' . $card->getFrom()
                    . ' to ' . $card->getTo()
                    . '. ' . $card->getInfo();
            }
        }
        $output[] = 'You have arrived at your final destination.';

        return $output;
    }
}


src/BoardingCard.php
<?php

namespace Skocic\TripSorter;

class BoardingCard
{
    /** @var string */
    private $from;
    /** @var string */
    private $to;
    /** @var string */
    private $type;
    /** @var string */
    private $number;
    /** @var string */
    private $info;

    /**
     * BoardingCard constructor.
     *
     * @param array $card
     */
    public function __construct(array $card = [])
    {
        $this->from = isset($card['from']) ? $card['from'] : '';
        $this->to = isset($card['to']) ? $card['to'] : '';
        $this->type = isset($card['type']) ? $card['type'] : '';
        $this->number = isset($card['number']) ? $card['number'] : '';
        $this->info = isset($card['info']) ? $card['info'] : '';
    }

    /**
     * Convert to array.
     *
     * @return array
     */
    public function toArray(): array
    {
        return [
            'from' => $this->getFrom(),
            'to' => $this->getTo(),
            'type' => $this->getType(),
            'number' => $this->getNumber(),
            'info' => $this->getInfo(),
        ];
    }

    /**
     * @return string
     */
    public function getFrom(): string
    {
        return $this->from;
    }

    /**
     * @return string
     */
    public function getTo(): string
    {
        return $this->to;
    }

    /**
     * @return string
     */
    public function getType(): string
    {
        return $this->type;
    }

    /**
     * @return string
     */
    public function getNumber(): string
    {
        return $this->number;
    }

    /**
     * @return string
     */
    public function getInfo(): string
    {
        return $this->info;
    }
}


src/TripSorter.php
<?php

namespace Skocic\TripSorter;

class TripSorter
{
    /**
     * Sort an array of trip cards.
     *
     * Example:
     * $cards = [
     *   ['from' => 'Madrid', 'to' => 'Barcelona', 'type' => 'train', 'number' => '78A', 'info' => 'Sit in seat 45B.'],
     *   ['from' => 'Barcelona', 'to' => 'Gerona Airport', 'type' => 'airport bus', 'number' => '', 'info' => 'No seat assignment.'],
     *   ['from' => 'Gerona Airport', 'to' => 'Stockholm', 'type' => 'flight', 'number' => 'SK455', 'seat' => '3A', 'info' => 'Gate 45B, seat 3A. Baggage drop at ticket counter 344.'],
     *   ['from' => 'Stockholm', 'to' => 'New York JFK', 'type' => 'flight', 'number' => 'SK455', 'seat' => '7B', 'info' => 'Gate 22. Baggage will we automatically transferred from your last leg.'],
     * ];
     * $result = TripSorter::sort($cards);
     *
     * returns
     * [
     *     'Take train 78A from Madrid to Barcelona. Sit in seat 45B.',
     *     'Take airport bus  from Barcelona to Gerona Airport. No seat assignment.',
     *     'From Gerona Airport, take flight SK455 to Stockholm. Gate 45B, seat 3A. Baggage drop at ticket counter 344.',
     *     'From Stockholm, take flight SK455 to New York JFK. Gate 22. Baggage will we automatically transferred from your last leg.',
     *     'You have arrived at your final destination.',
     * ];
     *
     * @param BoardingCard[] $cards
     * @return BoardingCard[]
     */
    public function sort(array $cards = []): array
    {
        $count = 0;
        $fromCity = [];
        $toCity = [];
        foreach ($cards as $card) {
               $toCity[$card->getTo()] = $count;
            $fromCity[$card->getFrom()] = $count++;
        }
        if (empty($fromCity)) {
            return [];
        }
        $diff = array_keys(array_diff_key($fromCity, $toCity));
        if (count($diff) !== 1) {
            throw new TripException('There should be one starting point');
        }
        $city = $diff[0];
        $sorted = [];
        while (isset($fromCity[$city])) {
            $index = $fromCity[$city];
            $sorted[] = $cards[$index];
            $city = $cards[$index]->getTo(); // Barcelona
        }

        return $sorted;
    }
}
